url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/gui/quantum/Item.java

```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+ * Copyright (c) 2013 Martin Marinov.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Public License v3.0
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/gpl.html
+ * 
+ * Contributors:
+ *     Martin - initial API and implementation
+ ******************************************************************************/
+package martin.gui.quantum;
+
+import java.awt.Cursor;
+import java.awt.Graphics2D;
+import java.awt.image.BufferedImage;
+import java.io.IOException;
+import java.util.HashSet;
+
+import javax.imageio.ImageIO;
+
+/**
+ * The {@link Item} serves two purposes. <br/>
+ * <ul>
+ * <li>Creating of similar {@link Item}s by interactively using the mouse and clicking onto a {@link Visualizer}.
+ * See {@link #mouseClick(Graphics2D, int, int, Visualizer)}, {@link #mouseDrag(Graphics2D, int, int, Visualizer)},
+ * {@link #mouseMove(Graphics2D, int, int, Visualizer)} and {@link #getCursor()} for information how this mode works.</li>
+ * <li>For actually rendering the {@link Item} on the screen one it has been created. See {@link #renderInstance(Graphics2D)}</li>
+ * </ul>
+ * Each item should implement its own {@link #getIconFileName()} in order for the interaction to work properly.
+ * 
+ * @author Martin Marinov
+ *
+ */
+public abstract class Item  {
+	
+	protected BufferedImage icon;
+	
+	/**
+	 * Render the item onto the visualizer. This is for active qubits, not for their creation.
+	 * @param g the graphics that is used to draw ontop of
+	 */
+	abstract void renderInstance(final Graphics2D g, Visualizer vis);
+	
+	/**
+	 * @param x
+	 * @param y
+	 * @return true if current {@link Item} contains points x and y on screen
+	 */
+	boolean isMouseOntop(int x, int y) {return false;};
+	
+	/**
+	 * Return the file name that would be used as an icon
+	 * @return
+	 */
+	abstract String getIconFileName();
+	
+	/**
+	 * Someone requires the item to move with a certain amount
+	 * @param dx the amount in the x direction that the mouse moved
+	 * @param dy the amount in the y direction that the mouse moved
+	 * @param mx the x position of the mouse
+	 * @param my the y position of the mouse
+	 * @param g
+	 * @param vis
+	 */
+	void moveWith(int dx, int dy, int mx, int my, final Graphics2D g, Visualizer vis) {};
+	
+	/**
+	 * Override if you don't want to use icons from files
+	 * @return
+	 */
+	BufferedImage getIcon() {
+		try {
+			return icon = ImageIO.read(this.getClass().getResourceAsStream(getIconFileName()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getResourceAsStream*. Currently, there are execution paths that do not contain closure statements, for example, when ImageIO.read throws an exception. Ensure the object returned by *getResourceAsStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).